torch/csrc/jit/codegen/cuda/iter_visitor.cpp

```diff
@@ -467,6 +467,84 @@ void BackwardVisitor::traverseFrom(
   }
 }
 
+/* SUBGRAPH VISITOR */
+
+std::vector<Statement*> SubgraphVisitor::next(Statement* stmt) {
+  if (stmt->isVal()) {
+    return next(stmt->as<Val>());
+  } else {
+    return next(stmt->as<Expr>());
+  }
+}
+
+std::vector<Statement*> SubgraphVisitor::next(Val* v) {
+  std::vector<Statement*> next_stmts;
+  FusionGuard::getCurFusion()->assertInContainer(v, "Cannot traverse val, ");
+  next_stmts.reserve(v->uses().size() + 1);
+  next_stmts.insert(next_stmts.end(), v->uses().begin(), v->uses().end());
+  if (v->definition() != nullptr) {
+    next_stmts.emplace_back(v->definition());
+  }
+  return next_stmts;
+}
+
+std::vector<Statement*> SubgraphVisitor::next(Expr* expr) {
+  std::vector<Statement*> next_stmts;
+  FusionGuard::getCurFusion()->assertInContainer(
+      expr, "Cannot traverse expr, ");
+  next_stmts.reserve(expr->inputs().size() + expr->outputs().size());
+  next_stmts.insert(
+      next_stmts.end(), expr->inputs().begin(), expr->inputs().end());
+  next_stmts.insert(
+      next_stmts.end(), expr->outputs().begin(), expr->outputs().end());
+  return next_stmts;
+}
+
+void SubgraphVisitor::traverseFrom(
+    Fusion* fusion,
+    const std::vector<Val*>& from) {
+  FusionGuard fg(fusion);
+  to_visit.clear();
+  queued.clear();
+  to_visit.insert(to_visit.end(), from.begin(), from.end());
+  std::copy(from.begin(), from.end(), std::inserter(queued, queued.end()));
+  while (!to_visit.empty()) {
+    auto node = to_visit.back();
+    to_visit.pop_back();
+    handle(node);
+    for (auto n : next(node)) {
+      if (queued.find(n) == queued.end()) {
+        to_visit.emplace_back(n);
+        queued.insert(n);
+      }
+    }
+  }
+}
+
+namespace {
+
+struct SubgraphInputs : public SubgraphVisitor {
+  std::vector<Val*> graph_inputs;
+
+ protected:
+  virtual void handle(Val* val) override {
+    if (val->definition() == nullptr) {
+      graph_inputs.emplace_back(val);
+    }
+  }
+};
+
+} // namespace
+
+std::vector<Val*> SubgraphVisitor::getAllInputs(const std::vector<Val*>& vals) {
```

Comment:
This feels overly complicated to me. `SuggraphInputs` inherits from `SubgraphVisitor`, and this method of `SubgraphVisitor` directly uses its child class. Could the class inheritance be simplified?